File: Graphite/Declarations.cs
namespace Graphite
{
    public record LibraryDeclaration(string name, ModuleDeclaration[] modules) { }
    public record ModuleDeclaration(string name, MethodDeclaration[] methods) { }
    public record MethodDeclaration(string name, MethodParameterDeclaration[] parameters, IStatement[] statements) : IMethodDeclaration { }
    public record MethodParameterDeclaration(string name, Type type) { }

    public record ExternalLibraryDeclaration(string name, ExternalModuleDeclaration[] modules) { }
    public record ExternalModuleDeclaration(string name, ExternalMethodDeclaration[] methods) { }
    public record ExternalMethodDeclaration(string name, MethodParameterDeclaration[] parameters) : IMethodDeclaration { }

    public interface IMethodDeclaration
    {
        string name { get; }
    }

}

File: Graphite.Core/Declarations.fs
namespace Graphite.Core

type PrimitiveType = String
type Type = PrimitiveType

type MethodParameterDeclaration = { name: string; ``type``: Type }

type InternalMethodDeclaration = { name: string; parameters: array<MethodParameterDeclaration>; (*statements: array<IStatement>*) }
type InternalModuleDeclaration = { name: string; methods: array<InternalMethodDeclaration> }
type InternalLibraryDeclaration = { name: string; modules: array<InternalModuleDeclaration> }

type ExternalMethodDeclaration = { name: string; parameters: array<MethodParameterDeclaration> }
type ExternalModuleDeclaration = { name: string; methods: array<ExternalMethodDeclaration> }
type ExternalLibraryDeclaration = { name: string; modules: array<ExternalModuleDeclaration> }

type MethodDeclaration = 
    | Internal of InternalMethodDeclaration 
    | External of ExternalMethodDeclaration  
    
type ModuleDeclaration = 
    | Internal of InternalModuleDeclaration 
    | External of ExternalModuleDeclaration      
    
type LibraryDeclaration = 
    | Internal of InternalLibraryDeclaration 
    | External of ExternalLibraryDeclaration

type StringLiteral = string
type Value = StringLiteral

type Argument = { parameter: MethodParameterDeclaration; value: Value }
type MethodInvocation = { method: MethodDeclaration; arguments: array<Argument> }

type Statement =
    | Invocation of MethodInvocation

type Program = { libraries:array<ExternalLibraryDeclaration>; entrypoint: array<Statement> }